hs_core/hydroshare/features.py

```diff
@@ -0,0 +1,475 @@
+from django.contrib.auth.models import User, Group
+from hs_access_control.models import PrivilegeCodes, GroupAccess, GroupResourcePrivilege
+from hs_core.models import BaseResource
+from hs_core.search_indexes import BaseResourceIndex
+from hs_tracking.models import Variable
+import re
+
+
+class Features(object):
+    """
+    Features appropriate for machine learning analysis of our databases.
+    """
+
+    @staticmethod
+    def resource_owners():
```

Comment:
These methods could use some optimizations and won't scale well.  All of this logic may be done in the db, which will be significantly faster than loading every user and every resources in python objects when you only need the username and resource short_id, something like this.  

```
@staticmethod
def group_owners():
	""" return (username, group_name) tuples representing group ownership """
	return group_privileges(PrivilegeCodes.OWNER)

@staticmethod
def group_privileges(privilege_code):
	user_ids = User.objects.all().values_list("pk", flat=True)
	groups = Group.objects.filter(g2ugp__user__id__in=user_ids, g2ugp__privilege=privilege_code)
	records = groups.values_list("g2ugp__user__username", "name")
	sorted_records = records.order_by("g2ugp__user")
	return [c for c in sorted_records.all()]
```

Reply:
I will discuss the code in the file with Alva. It looks like functions in this file is not being used in the current system. We are thinking about whether we are going to remove code from this file for this PR.